core/src/main/java/io/questdb/cairo/map/Unordered2Map.java

```diff
@@ -76,59 +76,64 @@ public Unordered2Map(
+                if (size > 0) {
+                    keySize += size;
+                } else {
+                    keySize = -1;
+                    break;
+                }
+            }
+            if (keySize <= 0 || keySize > KEY_SIZE) {
+                throw CairoException.nonCritical().put("unexpected key size: ").put(keySize);
+            }
+
+            long valueOffset = 0;
+            long[] valueOffsets = null;
+            long valueSize = 0;
+            if (valueTypes != null) {
+                int valueColumnCount = valueTypes.getColumnCount();
+                valueOffsets = new long[valueColumnCount];
+
+                for (int i = 0; i < valueColumnCount; i++) {
+                    valueOffsets[i] = valueOffset;
+                    final int columnType = valueTypes.getColumnType(i);
+                    final int size = ColumnType.sizeOf(columnType);
+                    if (size <= 0) {
+                        throw CairoException.nonCritical().put("value type is not supported: ").put(ColumnType.nameOf(columnType));
+                    }
+                    valueOffset += size;
+                    valueSize += size;
                 }
-                valueOffset += size;
-                valueSize += size;
             }
-        }
 
-        this.entrySize = Bytes.align2b(KEY_SIZE + valueSize);
+            this.entrySize = Bytes.align2b(KEY_SIZE + valueSize);
 
-        final long sizeBytes = entrySize * TABLE_SIZE;
-        memStart = Unsafe.malloc(sizeBytes, memoryTag);
-        Vect.memset(memStart, sizeBytes, 0);
-        memLimit = memStart + sizeBytes;
-        keyMemStart = Unsafe.malloc(KEY_SIZE, memoryTag);
-        Unsafe.getUnsafe().putShort(keyMemStart, (short) 0);
+            final long sizeBytes = entrySize * TABLE_SIZE;
+            memStart = Unsafe.malloc(sizeBytes, memoryTag);
+            Vect.memset(memStart, sizeBytes, 0);
+            memLimit = memStart + sizeBytes;
+            keyMemStart = Unsafe.malloc(KEY_SIZE, memoryTag);
+            Unsafe.getUnsafe().putShort(keyMemStart, (short) 0);
 
-        value = new Unordered2MapValue(valueSize, valueOffsets);
-        value2 = new Unordered2MapValue(valueSize, valueOffsets);
-        value3 = new Unordered2MapValue(valueSize, valueOffsets);
+            value = new Unordered2MapValue(valueSize, valueOffsets);
+            value2 = new Unordered2MapValue(valueSize, valueOffsets);
+            value3 = new Unordered2MapValue(valueSize, valueOffsets);
 
-        record = new Unordered2MapRecord(valueSize, valueOffsets, value, keyTypes, valueTypes);
-        cursor = new Unordered2MapCursor(record, this);
-        key = new Key();
+            record = new Unordered2MapRecord(valueSize, valueOffsets, value, keyTypes, valueTypes);
+            cursor = new Unordered2MapCursor(record, this);
+            key = new Key();
+        } catch (Throwable th) {
+            close();
```

Comment:
there is a problem if we manage to alloc `memstart` but fail to alloc `keyMemStart` we will be freeing 0-pointer

Reply:
Freeing a null-pointer is fine, it checks that and returns.

```java
    public static long free(long ptr, long size, int memoryTag) {
        if (ptr != 0) {
            Unsafe.getUnsafe().freeMemory(ptr);
            FREE_COUNT.incrementAndGet();
            recordMemAlloc(-size, memoryTag);
        }
        return 0;
    }
```